[Skia] Add initial support for color spaces

Scony opened the pull request:
#### 3466e3a573ebcaa89a49e3f0649627c1409435f9
<pre>
[Skia] Add initial support for color spaces
<a href="https://bugs.webkit.org/show_bug.cgi?id=272453">https://bugs.webkit.org/show_bug.cgi?id=272453</a>

Reviewed by Carlos Garcia Campos.

With this change, the colorspaces are now enabled in all the codepaths requiring them.
This change leaves some codepaths unimplemented - to be implemented in followup PRs.
All the SVG filters are working now as good as they worked with cairo.

* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::convertAndResolveColorComponents):
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::SRGB):
(WebCore::DestinationColorSpace::LinearSRGB):
(WebCore::DestinationColorSpace::DisplayP3):
(WebCore::operator==):
(WebCore::DestinationColorSpace::asRGB const):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
(WebCore::DestinationColorSpace::platformColorSpace const):
* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::convertImagePixelsSkia):
(WebCore::convertImagePixels):
* Source/WebCore/platform/graphics/PlatformColorSpace.h:
* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
* Source/WebCore/platform/graphics/filters/FEFlood.cpp:
(WebCore::FEFlood::create):
* Source/WebCore/platform/graphics/filters/FEFlood.h:
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::requiresPixelBufferColorSpaceConversion const):
(WebCore::FilterImage::transformToColorSpace):
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
(WebCore::ImageGStreamer::ImageGStreamer):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::UnacceleratedBuffer::UnacceleratedBuffer):
* Source/WebCore/platform/graphics/skia/ColorSpaceSkia.cpp: Copied from Source/WebCore/platform/image-decoders/skia/ImageBackingStoreSkia.cpp.
(WebCore::sRGBColorSpaceRef):
(WebCore::linearSRGBColorSpaceRef):
(WebCore::displayP3ColorSpaceRef):
* Source/WebCore/platform/graphics/skia/ColorSpaceSkia.h: Copied from Source/WebCore/platform/image-decoders/skia/ImageBackingStoreSkia.cpp.
* Source/WebCore/platform/graphics/skia/GraphicsContextGLSkia.cpp:
(WebCore::GraphicsContextGL::createNativeImageFromPixelBuffer):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::GraphicsContextSkia):
(WebCore::GraphicsContextSkia::colorSpace const):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::create):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.cpp:
(WebCore::ImageBufferSkiaUnacceleratedBackend::create):
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::colorSpace const):
* Source/WebCore/platform/graphics/skia/ShareableBitmapSkia.cpp:
(WebCore::ShareableBitmapConfiguration::calculateBytesPerPixel):
(WebCore::ShareableBitmapConfiguration::calculateBytesPerRow):
* Source/WebCore/platform/graphics/skia/SkiaAcceleratedBufferPool.cpp:
(WebCore::SkiaAcceleratedBufferPool::createAcceleratedBuffer):
* Source/WebCore/platform/image-decoders/skia/ImageBackingStoreSkia.cpp:
(WebCore::ImageBackingStore::image const):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyMask):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceMasker.cpp:
(WebCore::LegacyRenderSVGResourceMasker::applyResource):
(WebCore::LegacyRenderSVGResourceMasker::drawContentIntoMaskImage):
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/skia/CoreIPCSkColorSpace.h: Copied from Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp.
(WebKit::CoreIPCSkColorSpace::CoreIPCSkColorSpace):
(WebKit::CoreIPCSkColorSpace::skColorSpace const):
(WebKit::CoreIPCSkColorSpace::dataReference const):
* Source/WebKit/Shared/skia/CoreIPCSkColorSpace.serialization.in: Added.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
(IPC::ArgumentCoder&lt;sk_sp&lt;SkColorSpace&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;sk_sp&lt;SkColorSpace&gt;&gt;::decode):
* Source/WebKit/UIProcess/cairo/BackingStoreCairo.cpp:
(WebKit::BackingStore::incorporateUpdate):
* Tools/WebKitTestRunner/gtk/PlatformWebViewGtk.cpp:
(WTR::PlatformWebView::windowSnapshotImage):
* Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.cpp:
(WTR::PlatformWebViewClientWPE::snapshot):
* Tools/wpe/backends/fdo/HeadlessViewBackendFdo.cpp:
(WPEToolingBackends::HeadlessViewBackend::HeadlessViewBackend):
(WPEToolingBackends::HeadlessViewBackend::updateSnapshot):

Canonical link: <a href="https://commits.webkit.org/277802@main">https://commits.webkit.org/277802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fce8798ee973bb3771d94bf7b00a61e6d88a56c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48567 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27778 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51528 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51254 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44632 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50872 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33714 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25307 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39711 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49149 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25445 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41899 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20808 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22922 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43073 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6623 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44860 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43561 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53161 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23613 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19913 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47004 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24879 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42105 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45928 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25683 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6931 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24601 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

TingPing commented on Source/WebCore/platform/graphics/PlatformColorSpace.h:
```diff
@@ -39,6 +41,26 @@ namespace WebCore {
+class PlatformColorSpace {
+public:
+    PlatformColorSpace(int serializedColorSpace) { deserialize(serializedColorSpace); }
+    PlatformColorSpace(sk_sp<SkColorSpace> colorSpace)
```
```suggestion
    explicit PlatformColorSpace(int serializedColorSpace) { deserialize(serializedColorSpace); }
    explicit PlatformColorSpace(sk_sp<SkColorSpace> colorSpace)
```

Scony replied:
While making `PlatformColorSpace(int serializedColorSpace)` `explicit` makes a lot of sense, in case of `PlatformColorSpace(sk_sp<SkColorSpace> colorSpace)` I think it's not worth it as in many cases we're actually leveraging implicit conversions to avoid constructions like `DestinationColorSpace(PlatformColorSpace(skspcs))`

carlosgcampos commented on Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
```diff
@@ -137,6 +139,50 @@ static void convertImagePixelsAccelerated(const ConstPixelBufferConversionView&
+{
+    auto toSkiaColorType = [](const PixelFormat& pixelFormat) {
+        switch (pixelFormat) {
+        case PixelFormat::RGBA8: return SkColorType::kRGBA_8888_SkColorType;
```
```suggestion
        case PixelFormat::RGBA8: 
            return SkColorType::kRGBA_8888_SkColorType;
```

carlosgcampos commented on Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
```diff
@@ -137,6 +139,50 @@ static void convertImagePixelsAccelerated(const ConstPixelBufferConversionView&
+    auto toSkiaColorType = [](const PixelFormat& pixelFormat) {
+        switch (pixelFormat) {
+        case PixelFormat::RGBA8: return SkColorType::kRGBA_8888_SkColorType;
+        case PixelFormat::BGRA8: return SkColorType::kBGRA_8888_SkColorType;
```
```suggestion
        case PixelFormat::BGRA8: 
            return SkColorType::kBGRA_8888_SkColorType;
```

carlosgcampos commented on Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
```diff
@@ -137,6 +139,50 @@ static void convertImagePixelsAccelerated(const ConstPixelBufferConversionView&
+    };
+    auto toSkiaAlphaType = [](const AlphaPremultiplication& alphaFormat) {
+        switch (alphaFormat) {
+        case AlphaPremultiplication::Premultiplied: return SkAlphaType::kPremul_SkAlphaType;
```
```suggestion
        case AlphaPremultiplication::Premultiplied: 
            return SkAlphaType::kPremul_SkAlphaType;
```

carlosgcampos commented on Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
```diff
@@ -137,6 +139,50 @@ static void convertImagePixelsAccelerated(const ConstPixelBufferConversionView&
+    auto toSkiaAlphaType = [](const AlphaPremultiplication& alphaFormat) {
+        switch (alphaFormat) {
+        case AlphaPremultiplication::Premultiplied: return SkAlphaType::kPremul_SkAlphaType;
+        case AlphaPremultiplication::Unpremultiplied: return SkAlphaType::kUnpremul_SkAlphaType;
```
```suggestion
        case AlphaPremultiplication::Unpremultiplied: 
            return SkAlphaType::kUnpremul_SkAlphaType;
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ColorSpaceSkia.cpp:
```diff
@@ -0,0 +1,67 @@
+void PlatformColorSpace::deserialize(int)
+{
+    notImplemented();
+}
```
I guess we can add these when needed, no? because this looks specific to skia.

Scony replied:
I tried pretty hard to not include that but I couldn't find a way to avoid serialization in `Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in` for now. I have an impression that this is mandatory - e.g. in cairo case we're using a value of `enum class Name : uint8_t` and in case of CG we're using RefPtr's managed object pointer. So having dummy int is a bare minimum I could figure out. In the future we will use an array of `uint8_t`.

carlosgcampos replied:
Can we implement it instead then? It's difficult for me to review it without knowing what serialize and deserialize are going to do.

Scony replied:
Ok

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -480,11 +498,22 @@ SkPaint GraphicsContextSkia::createStrokePaint() const
+void GraphicsContextSkia::setupColor(SkPaint& paint, SkColor color) const
+{
+    paint.setColor4f(SkColor4f::FromColor(color), colorSpace().platformColorSpace().get());
+}
```
I think it's better to do this in every caller, it will be a long line, but you don't have to go to the helper function to know what setting up color means

carlosgcampos replied:
hmm, I'm, not sure I understand this correctly either, I think the color space parameter is to specify the color space of the color we are setting, not the destination color space. Or am I misunderstanding what `colorSpace()` is?

Scony replied:
You're correct. And since `SkPaint` assumes colors without specified color space are in `sRGB`, we need to make sure the color will be interpreted as if it was in destination color space so that no color conversions are done.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -480,11 +498,22 @@ SkPaint GraphicsContextSkia::createStrokePaint() const
+void GraphicsContextSkia::setupShader(SkPaint& paint, sk_sp<SkShader> shader) const
+{
+    if (shader)
+        paint.setShader(shader->makeWithWorkingColorSpace(colorSpace().platformColorSpace()));
```
Same here. I think this is a different behavior, in case of null shader we don't set it. I guess it's not a problem because we usually call this for a newly created SkPaint, so it's already null

Scony replied:
The problem I see here is that every caller such as `paint.setShader(fillPattern->createPlatformPattern({ }, toSkSamplingOptions(imageInterpolationQuality())))` that doesn't have to check if shader is null or not would have to be accompanied by an `if` statement now so that we can safely `->makeWithWorkingColorSpace`.
Alternatively we could make sure that every producer of `sk_sp<SkShader>` returns shader that is not null, but that may lead to crashes if someone adds a new, unchecked producer.

carlosgcampos replied:
The color space we are using here is the graphics context one, right? and we got that one from the SkCanvas, so it's the current destination color space. Doesn't shaders use the destination color space by default?

Scony replied:
I've checked and you're right - `By default, all shaders operate in the destination (surface) color space` - so it looks like in case of shaders it's different than in case of `SkPaint`'s colors. I'll remove the above then.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -55,11 +56,23 @@
+            colorSpace = surface->imageInfo().refColorSpace();
+    }
+    // No color space means the default - SRGB.
+    return colorSpace != nullptr ? DestinationColorSpace(colorSpace) : DestinationColorSpace::SRGB();
```
Don't compare to nullptr, use !colorSpace instead

carlosgcampos commented on Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
```diff
@@ -61,7 +61,12 @@ bool PlatformImageNativeImageBackend::hasAlpha() const
 {
-    notImplemented();
+    auto colorSpace = platformImage()->refColorSpace();
+    if (colorSpace == nullptr)
```
```suggestion
    if (!colorSpace)
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
```diff
@@ -61,7 +61,12 @@ bool PlatformImageNativeImageBackend::hasAlpha() const
-    notImplemented();
+    auto colorSpace = platformImage()->refColorSpace();
+    if (!colorSpace)
+        colorSpace = platformImage()->imageInfo().refColorSpace();
```
This is the same, `SkImage::refColorSpace()` returns `fInfo.refColorSpace()`, and `SkImage::imageInfo()` returns `fInfo`.

carlosgcampos commented on Source/WebCore/platform/graphics/PlatformColorSpace.h:
```diff
@@ -39,6 +42,26 @@ namespace WebCore {
+
+class PlatformColorSpace {
+public:
+    explicit PlatformColorSpace(Vector<uint8_t> serializedColorSpace) { deserialize(serializedColorSpace); }
```
So, if I understand this correctly now, we need this wrapper class instead of using PlatformColorSpace as sk_sp<SkColorSpace> directly only for the serialization, right? I don't know much about the new serializers, but would it be possible to add a CoreIPC wrapper of SkColorSpace to serialize/deserialize it and use PlatformColorSpace as sk_sp<SkColorSpace> like CG does for CGColorSpaceRef?

Scony replied:
Correct - if it wasn't for serialization we could just use `sk_sp<SkColorSpace>` as `PlatformColorSpace` and `PlatformColorSpaceValue`.
As for `CoreIPC wrapper` I'm not sure what you mean, but if it's about providing custom serializer, we'd have to figure out how to avoid generating default ones in `WebCoreArgumentCoders.serialization.in` as I've found it not trivial last time I've checked.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -55,11 +56,23 @@
+    if (!colorSpace) {
+        if (auto surface = canvas.getSurface(); surface != nullptr)
+            colorSpace = surface->imageInfo().refColorSpace();
+    }
```
In this case I'm not sure, but I suspect SkCanvas::imageInfo() returns the same value as its surface.

Scony replied:
In that particular case I've actually encountered the case where `canvas.imageInfo().refColorSpace()` is null and yet `surface->imageInfo().refColorSpace()` is not

Scony replied:
I've added extra comment here:

> Skia sometimes fail to propagate color space info correctly, so we need to fetch few places.

It turns out that on some occasions skia won't propagate the color space information to `SkCanvas`. Actually this was the problem that was leading to lack of conversion when performing `drawImageRect()` on canvas. I've fixed that by specifying color space explicitly in https://github.com/WebKit/WebKit/pull/27079/files#diff-e07ba6513cad2cefec7a94fd9bdf37e71ea228e3f3570c9be1d4aed9dbadd877R82 so that when we're creating surface using e.g. `SkSurfaces::RenderTarget()` the info with non-null color space is used.

I've update other `SkImageInfo`'s constructing calls, but on some occasions `SkCanvas` still has image info out-of-sync with the surface. Looks like some kind of skia bug, so the above change will still be necessary.

carlosgcampos replied:
hmm, so even if explicitly adding SkColorSpace::MakeSRGB() doesn't cover all the cases, I think it's better to simply handle the null colorspace here and default to sRGB, no? We might add new `SkImageInfo::Make` calls and forget about the color space parameter.

Scony replied:
That would still break `SkCanvas::drawImageRect()`. As suggested offline, we will discuss this matter with skia devs directly.

Scony replied:
To sum up, the quote from skia dev:

> You should definitely be passing sRGB explicitly (not nullptr), when creating the surface. For sources (eg, images), nullptr is interpreted as sRGB. But for destinations (surfaces), nullptr actually means: "do no color management". This is an ugly legacy option

As for the above code for fetching color space info, again quote from skia dev:

> The discrepancy between the canvas and surface (assuming those are the same underlying surface) is definitely unexpected

So, I'll try to align our `SkImageInfo::Make` calls with what's been mentioned above.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -293,7 +311,7 @@ void GraphicsContextSkia::drawLine(const FloatPoint& point1, const FloatPoint& p
 
     SkPaint paint = createStrokePaint();
-    paint.setColor(SkColor(strokeColor().colorWithAlphaMultipliedBy(alpha())));
+    paint.setColor(strokeColor().colorWithAlphaMultipliedBy(alpha()), colorSpace().platformColorSpace().get());
```
I still think the color space parameters means the color space of the given color, not the destination color space. When converting Color to Skcolor we always get the Color as SRGBA, so I don't think we need to pass any color space to SkPaint::setColor

Scony replied:
You're correct color space parameters mean the color space of the given color. Also, as you mentioned - when converting Color to SkColor we always get the Color as SRGBA. The thing is, the callers of GraphicsContext assume that the color they provide will be exactly what will be written to the destination. So if some caller (let's say feFlood filter) uses GraphicsContext that has linear SRGB destination and calls fillRect with color `(0.501961, 0.501961, 0.501961, 1.000000)` it assumes the pixels will have values `(0.501961, 0.501961, 0.501961, 1.000000)`. Without the above parameter, the color would be treated as SRGB and therefore while filling rect it would be converted to linear SRGB so the pixels would be filled with color `(0.215686, 0.215686, 0.215686, 1.000000)`

Scony replied:
Actually, I've started checking callers on why they _"assume that the color they provide should be exactly what will be written to the destination"_ and I've discovered that they actually don't assume that and the problem is elsewhere - I'll update PR in few h.

Scony replied:
Done, see https://github.com/WebKit/WebKit/pull/27079#discussion_r1565786804

philn commented on Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
```diff
@@ -91,7 +91,7 @@ ImageGStreamer::ImageGStreamer(GRefPtr<GstSample>&& sample)
     }
 
-    auto imageInfo = SkImageInfo::Make(videoFrame->width(), videoFrame->height(), colorType, alphaType);
+    auto imageInfo = SkImageInfo::Make(videoFrame->width(), videoFrame->height(), colorType, alphaType, SkColorSpace::MakeSRGB());
```
Here instead of hardcoding sRGB, maybe get the colorimetry info from the GstVideoInfo. There is a `PlatformVideoColorSpace videoColorSpaceFromInfo(const GstVideoInfo&);` function in GStreamerCommon. Then somehow map the PlatformVideoColorSpace to the skia equivalent.

Scony replied:
I've checked and that conversion does not look trivial, therefore I've created a simple, not-implemented converter. If you don't mind I'll leave that converter for followup PR.

carlosgcampos commented on Source/WebCore/platform/graphics/PlatformColorSpace.h:
```diff
@@ -39,6 +41,11 @@ namespace WebCore {
+#elif USE(SKIA)
+
+using PlatformColorSpace = sk_sp<SkColorSpace>;
+using PlatformColorSpaceValue = sk_sp<SkColorSpace>;
```
I guess this would be `SkColorSpace*`, or do we want to use the smart pointer as value too?

Scony replied:
Actually we must use `sk_sp<SkColorSpace>` as in many cases we need to use `sk_sp<SkColorSpace>` given `PlatformColorSpaceValue`

carlosgcampos commented on Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
```diff
@@ -33,6 +33,8 @@
 #if USE(ACCELERATE) && USE(CG)
 #include <Accelerate/Accelerate.h>
+#elif USE(SKIA)
+#include <skia/core/SkPixmap.h>
```
```suggestion
IGNORE_CLANG_WARNINGS_BEGIN("cast-align")
#include <skia/core/SkPixmap.h>
IGNORE_CLANG_WARNINGS_END
```

carlosgcampos commented on Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
```diff
@@ -25,9 +25,11 @@
 
 #include "config.h"
 #include "WebCoreArgumentCoders.h"
+#include "StreamConnectionEncoder.h"
```
This should be moved inside the `#if USE(SKIA)` block

Scony committed:
[Skia] Add initial support for color spaces
https://bugs.webkit.org/show_bug.cgi?id=272453

Reviewed by Carlos Garcia Campos.

With this change, the colorspaces are now enabled in all the codepaths requiring them.
This change leaves some codepaths unimplemented - to be implemented in followup PRs.
All the SVG filters are working now as good as they worked with cairo.

* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::convertAndResolveColorComponents):
* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::SRGB):
(WebCore::DestinationColorSpace::LinearSRGB):
(WebCore::DestinationColorSpace::DisplayP3):
(WebCore::operator==):
(WebCore::DestinationColorSpace::asRGB const):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
(WebCore::DestinationColorSpace::platformColorSpace const):
* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::convertImagePixelsSkia):
(WebCore::convertImagePixels):
* Source/WebCore/platform/graphics/PlatformColorSpace.h:
* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
* Source/WebCore/platform/graphics/filters/FEFlood.cpp:
(WebCore::FEFlood::create):
* Source/WebCore/platform/graphics/filters/FEFlood.h:
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::requiresPixelBufferColorSpaceConversion const):
(WebCore::FilterImage::transformToColorSpace):
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
(WebCore::ImageGStreamer::ImageGStreamer):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::UnacceleratedBuffer::UnacceleratedBuffer):
* Source/WebCore/platform/graphics/skia/ColorSpaceSkia.cpp: Copied from Source/WebCore/platform/image-decoders/skia/ImageBackingStoreSkia.cpp.
(WebCore::sRGBColorSpaceRef):
(WebCore::linearSRGBColorSpaceRef):
(WebCore::displayP3ColorSpaceRef):
* Source/WebCore/platform/graphics/skia/ColorSpaceSkia.h: Copied from Source/WebCore/platform/image-decoders/skia/ImageBackingStoreSkia.cpp.
* Source/WebCore/platform/graphics/skia/GraphicsContextGLSkia.cpp:
(WebCore::GraphicsContextGL::createNativeImageFromPixelBuffer):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::GraphicsContextSkia):
(WebCore::GraphicsContextSkia::colorSpace const):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::create):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaUnacceleratedBackend.cpp:
(WebCore::ImageBufferSkiaUnacceleratedBackend::create):
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::colorSpace const):
* Source/WebCore/platform/graphics/skia/ShareableBitmapSkia.cpp:
(WebCore::ShareableBitmapConfiguration::calculateBytesPerPixel):
(WebCore::ShareableBitmapConfiguration::calculateBytesPerRow):
* Source/WebCore/platform/graphics/skia/SkiaAcceleratedBufferPool.cpp:
(WebCore::SkiaAcceleratedBufferPool::createAcceleratedBuffer):
* Source/WebCore/platform/image-decoders/skia/ImageBackingStoreSkia.cpp:
(WebCore::ImageBackingStore::image const):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyMask):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceMasker.cpp:
(WebCore::LegacyRenderSVGResourceMasker::applyResource):
(WebCore::LegacyRenderSVGResourceMasker::drawContentIntoMaskImage):
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/skia/CoreIPCSkColorSpace.h: Copied from Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp.
(WebKit::CoreIPCSkColorSpace::CoreIPCSkColorSpace):
(WebKit::CoreIPCSkColorSpace::skColorSpace const):
(WebKit::CoreIPCSkColorSpace::dataReference const):
* Source/WebKit/Shared/skia/CoreIPCSkColorSpace.serialization.in: Added.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
(IPC::ArgumentCoder<sk_sp<SkColorSpace>>::encode):
(IPC::ArgumentCoder<sk_sp<SkColorSpace>>::decode):
* Source/WebKit/UIProcess/cairo/BackingStoreCairo.cpp:
(WebKit::BackingStore::incorporateUpdate):
* Tools/WebKitTestRunner/gtk/PlatformWebViewGtk.cpp:
(WTR::PlatformWebView::windowSnapshotImage):
* Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.cpp:
(WTR::PlatformWebViewClientWPE::snapshot):
* Tools/wpe/backends/fdo/HeadlessViewBackendFdo.cpp:
(WPEToolingBackends::HeadlessViewBackend::HeadlessViewBackend):
(WPEToolingBackends::HeadlessViewBackend::updateSnapshot):

Canonical link: https://commits.webkit.org/277802@main

mcatanzaro commented on Source/WebKit/Shared/skia/CoreIPCSkColorSpace.h:
```diff
@@ -0,0 +1,61 @@
+
+class CoreIPCSkColorSpace {
+public:
+    CoreIPCSkColorSpace(sk_sp<SkColorSpace> skColorSpace)
+        : m_skColorSpace(WTFMove(skColorSpace))
+    {
+    }
+
+    CoreIPCSkColorSpace(std::span<const uint8_t> data)
```
These constructors should use `explicit`.
